packages/spatial/src/renderer/materials/MaterialComponent.ts

```diff
@@ -0,0 +1,166 @@
+WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
+specific language governing rights and limitations under the License.
+
+The Original Code is Ethereal Engine.
+
+The Original Developer is the Initial Developer. The Initial Developer of the
+Original Code is the Ethereal Engine team.
+
+All portions of the code written by the Ethereal Engine team are Copyright © 2021-2023 
+Ethereal Engine. All Rights Reserved.
+*/
+
+import { defineComponent, defineQuery } from '@etherealengine/ecs'
+import { Entity, EntityUUID, UndefinedEntity } from '@etherealengine/ecs/src/Entity'
+import { TransparencyDitheringPlugin } from '@etherealengine/engine/src/avatar/components/TransparencyDitheringComponent'
+import { removeMaterialInstance } from '@etherealengine/engine/src/scene/materials/functions/materialSourcingFunctions'
+import { PluginObjectType, PluginType } from '@etherealengine/spatial/src/common/functions/OnBeforeCompilePlugin'
+import { Material, Shader, WebGLRenderer } from 'three'
+import { NoiseOffsetPlugin } from './constants/plugins/NoiseOffsetPlugin'
+import MeshBasicMaterial from './prototypes/MeshBasicMaterial.mat'
+import MeshLambertMaterial from './prototypes/MeshLambertMaterial.mat'
+import MeshMatcapMaterial from './prototypes/MeshMatcapMaterial.mat'
+import MeshPhongMaterial from './prototypes/MeshPhongMaterial.mat'
+import MeshPhysicalMaterial from './prototypes/MeshPhysicalMaterial.mat'
+import MeshStandardMaterial from './prototypes/MeshStandardMaterial.mat'
+import MeshToonMaterial from './prototypes/MeshToonMaterial.mat'
+import { ShaderMaterial } from './prototypes/ShaderMaterial.mat'
+import { ShadowMaterial } from './prototypes/ShadowMaterial.mat'
+
+export type MaterialWithEntity = Material & { entity: Entity }
+
+export type MaterialStatus = 'LOADED' | 'MISSING' | 'UNLOADED'
+
+export type MaterialPrototypeConstructor = new (...args: any) => any
+export type MaterialPrototypeObjectConstructor = { [key: string]: MaterialPrototypeConstructor }
+export type MaterialPrototypeDefinition = {
+  prototypeId: string
+  prototypeConstructor: MaterialPrototypeConstructor
+  arguments: PrototypeArgument
+  onBeforeCompile?: (shader: Shader, renderer: WebGLRenderer) => void
+}
+
+export type PrototypeArgument = {
+  [_: string]: {
+    type: string
+    default: any
+    min?: number
+    max?: number
+    options?: any[]
+  }
+}
+
+export const materialSuffix = ' (Material)'
+
+export const MaterialPrototypeDefinitions = [
+  MeshBasicMaterial,
+  MeshStandardMaterial,
+  MeshMatcapMaterial,
+  MeshPhysicalMaterial,
+  MeshLambertMaterial,
+  MeshPhongMaterial,
+  MeshToonMaterial,
+  ShaderMaterial,
+  ShadowMaterial
+] as MaterialPrototypeDefinition[]
+
+export const MaterialPlugins = [NoiseOffsetPlugin, TransparencyDitheringPlugin]
+
+export enum MaterialComponents {
+  Instance,
+  State,
+  Prototype,
+  Plugin
+}
+
+export const materialByHash = {} as Record<string, EntityUUID>
```

Comment:
is there a reason these are not hyperflux states? should they be?

Reply:
My question as well, why should they be?

Reply:
in case we need to react to them? i guess it depends how/why they're used

Reply:
also allows for far easier inspection

Reply:
I agree with the inspection bit. we should be using the ecs for reacting to them though

Reply:
really I dont like having these to begin with but it just eliminates the need for excessive querying. the only one that's strictly necessary is the hashmap